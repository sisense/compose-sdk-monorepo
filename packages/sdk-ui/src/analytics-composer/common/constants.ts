export const CODE_TEMPLATES_INDENT = 6;
