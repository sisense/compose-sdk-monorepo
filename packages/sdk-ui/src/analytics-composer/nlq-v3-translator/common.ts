import {
  Attribute,
  createAttributeHelper,
  DateLevels,
  isDatetime,
  JaqlDataSourceForDto,
  MetadataTypes,
} from '@sisense/sdk-data';

import { NlqTranslationResult, NormalizedField, NormalizedTable } from '../types.js';
import { ATTRIBUTE_PREFIX } from './types.js';

function parseAttributeName(attributeName: string): {
  tableName: string;
  columnName: string;
  level?: string;
} {
  // Parse "DM.Commerce.Date.Years" -> tableName: "Commerce", columnName: "Date", level: "Years"
  // Parse "DM.Brand.Brand" -> tableName: "Brand", columnName: "Brand", level: undefined
  const parts = attributeName.split('.');
  if (parts.length < 3 || `${parts[0]}.` !== ATTRIBUTE_PREFIX) {
    throw new Error(
      `Invalid attribute name format: "${attributeName}". Expected format: "${ATTRIBUTE_PREFIX}TableName.ColumnName[.Level]"`,
    );
  }

  const tableName = parts[1];
  const columnName = parts[2];
  let level: string | undefined;

  // extract level if present (4th part)
  if (parts.length === 4) {
    level = parts[3];
  }

  return { tableName, columnName, level };
}

function findDataFieldAndLevel(
  attributeName: string,
  tables: NormalizedTable[],
): { field: NormalizedField; level?: string } {
  const { tableName, columnName, level } = parseAttributeName(attributeName);

  const table = tables.find((t) => t.name === tableName);
  if (!table) {
    throw new Error(`Table "${tableName}" not found in the data model`);
  }

  const field = table.columns.find((column) => column.name === columnName);
  if (!field) {
    throw new Error(`Column "${columnName}" not found in table "${tableName}"`);
  }

  if (level) {
    if (!isDatetime(field.dataType)) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is not a datetime column`,
      );
    }
    const validLevels = DateLevels.all;
    if (!validLevels.includes(level)) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Valid levels are: ${validLevels.join(
          ', ',
        )}`,
      );
    }

    // throw error for time levels on date only column
    const dateOnlyLevels = DateLevels.dateOnly;
    if (field.dataType === 'date' && !dateOnlyLevels.includes(level)) {
      throw new Error(
        `Invalid level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is only a date column, not a datetime column`,
      );
    }

    // throw error for date levels on time only column
    const timeOnlyLevels = DateLevels.timeOnly;
    if (field.dataType === 'time' && !timeOnlyLevels.includes(level)) {
      throw new Error(
        `Invalid level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is only a time column, not a date column`,
      );
    }
  }

  return { field, level };
}

/**
 * Validates that an attribute is compatible with text values
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a text type
 */
export function isTextAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.TextAttribute;
}

/**
 * Validates that an attribute is compatible with numeric values
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a numeric type
 */
export function isNumericAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.NumericAttribute;
}

/**
 * Validates that an attribute is a date/datetime type
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a date/datetime type
 */
export function isDateLevelAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.DateLevel;
}

/**
 * Validates that an attribute is compatible with text OR numeric values
 * Used for filters like equals, doesntEqual that accept both text and number
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is text or numeric type
 */
export function isTextOrNumericAttribute(attribute: Attribute): boolean {
  return isTextAttribute(attribute) || isNumericAttribute(attribute);
}

/**
 * Validates that an attribute is NOT a date type
 * Used for filters that should not work with date attributes
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is not a date type
 */
export function isNonDateLevelAttribute(attribute: Attribute): boolean {
  return !isDateLevelAttribute(attribute);
}

/**
 * Gets a display string for the attribute type
 *
 * @param attribute - The attribute to get the type display string for
 * @returns A human-readable string representing the attribute type
 */
export function getAttributeTypeDisplayString(attribute: Attribute): string {
  return attribute.type === MetadataTypes.TextAttribute
    ? 'text'
    : attribute.type === MetadataTypes.NumericAttribute
    ? 'numeric'
    : attribute.type === MetadataTypes.DateLevel
    ? 'date/datetime'
    : 'unknown';
}

export function createAttributeFromName(
  attributeName: string,
  dataSource: JaqlDataSourceForDto,
  tables: NormalizedTable[],
) {
  const { field, level } = findDataFieldAndLevel(attributeName, tables);

  return createAttributeHelper({
    expression: field.expression,
    dataType: field.dataType,
    granularity: level,
    format: undefined,
    sort: undefined,
    dataSource,
  });
}

export function getSuccessData<T>(result: NlqTranslationResult<T>): T {
  if (!result.success) throw new Error('Expected success result');
  return result.data;
}

export function getErrors<T>(result: NlqTranslationResult<T>): string[] {
  if (result.success) throw new Error('Expected error result');
  return result.errors.map((error) => error.message);
}
