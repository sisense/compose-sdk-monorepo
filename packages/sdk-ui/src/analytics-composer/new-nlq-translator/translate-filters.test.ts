import { describe, it, expect } from 'vitest';
import {
  translateFiltersFromJSONFunctionCall,
  translateFiltersJSON,
  translateHighlightsFromJSONFunctionCall,
} from './translate-filters.js';
import { isFilterRelations, withoutGuids, JSONArray } from '@sisense/sdk-data';
import {
  MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
  MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
} from '../__mocks__/mock-data-sources.js';
import { ParsedFunctionCall } from './common.js';
import { NlqTranslationResult } from '../types.js';

function getSuccessData<T>(result: NlqTranslationResult<T>): T {
  if (!result.success) throw new Error('Expected success result');
  return result.data;
}

function getErrors<T>(result: NlqTranslationResult<T>): string[] {
  if (result.success) throw new Error('Expected error result');
  return result.errors;
}

describe('translateFilters', () => {
  it('should translate filter relations', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.logic.and',
        args: [
          {
            function: 'filterFactory.logic.or',
            args: [
              {
                function: 'filterFactory.members',
                args: ['DM.Country.Country', ['United States', 'Canada']],
              },
              {
                function: 'filterFactory.members',
                args: ['DM.Brand.Brand', ['Brand A', 'Brand B']],
              },
            ],
          },
          {
            function: 'filterFactory.measureGreaterThan',
            args: [{ function: 'measureFactory.sum', args: ['DM.Commerce.Revenue'] }, 1000],
          },
        ],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(result.success).toBe(true);
    const data = getSuccessData(result);
    expect(isFilterRelations(data)).toBe(true);
    expect(withoutGuids(data)).toMatchSnapshot();
  });

  it('should translate filters', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
      { function: 'filterFactory.members', args: ['DM.Commerce.Date.Years', ['2024', '2025']] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );
    expect(result.success).toBe(true);
    const data = getSuccessData(result);
    expect(isFilterRelations(data)).toBe(false);
    expect(withoutGuids(data)).toMatchSnapshot();
  });

  it('should return error for nonexistent filterFactory function', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'filterFactory.nonexistentFunction', args: ['DM.Country.Country', 'test'] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Function "filterFactory.nonexistentFunction" not found in filterFactory or measureFactory.',
    );
  });

  it('should return error for nonexistent measureFactory function', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.measureGreaterThan',
        args: [
          { function: 'measureFactory.nonexistentFunction', args: ['DM.Commerce.Revenue'] },
          1000,
        ],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Function "measureFactory.nonexistentFunction" not found in filterFactory or measureFactory.',
    );
  });

  it('should return error for completely invalid factory name', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'invalidFactory.someFunction', args: ['DM.Country.Country', 'test'] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Function "invalidFactory.someFunction" not found in filterFactory or measureFactory.',
    );
  });

  it('should return error for function path without factory prefix', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'members', args: ['DM.Country.Country', ['United States']] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Function "members" not found in filterFactory or measureFactory.',
    );
  });

  it('should return error for attribute name with insufficient parts after DM', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'filterFactory.members', args: ['DM.Country', ['United States']] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Invalid attribute name format: "DM.Country". Expected format: "DM.TableName.ColumnName[.Level]".',
    );
  });

  it('should return error for attribute name with only DM prefix', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'filterFactory.members', args: ['DM.', ['United States']] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Invalid attribute name format: "DM.". Expected format: "DM.TableName.ColumnName[.Level]".',
    );
  });

  it('should return error for nonexistent table name', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.members',
        args: ['DM.NonExistentTable.SomeColumn', ['test']],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain('Table "NonExistentTable" not found in the data model.');
  });

  it('should return error for nonexistent column name', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.members',
        args: ['DM.Country.NonExistentColumn', ['test']],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Column "NonExistentColumn" not found in table "Country".',
    );
  });

  it('should return error for date level on non-datetime column', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.members',
        args: ['DM.Country.Country.Years', ['2024']],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Invalid date level "Years" in attribute "DM.Country.Country.Years". Column "Country.Country" is not a datetime column.',
    );
  });

  it('should return error for invalid date level on datetime column', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      {
        function: 'filterFactory.members',
        args: ['DM.Commerce.Date.InvalidLevel', ['2024']],
      },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain(
      'Invalid date level "InvalidLevel" in attribute "DM.Commerce.Date.InvalidLevel".',
    );
  });

  it('should return error when trying to use measure function as top-level filter', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue'] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain('Invalid filter JSON');
  });

  it('should return error when trying to use measure function with title as top-level filter', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'measureFactory.count', args: ['DM.Commerce.Revenue', 'Revenue Count'] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain('Invalid filter JSON');
  });

  it('should return error when trying to use average measure as top-level filter', () => {
    const mockFiltersJSON: ParsedFunctionCall[] = [
      { function: 'measureFactory.average', args: ['DM.Commerce.Revenue'] },
    ];

    const result = translateFiltersFromJSONFunctionCall(
      mockFiltersJSON,
      MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
      MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
    );

    expect(result.success).toBe(false);
    expect(getErrors(result)[0]).toContain('Invalid filter JSON');
  });

  describe('translateFiltersFromJSON', () => {
    it('should return empty array when filtersJSON is null', () => {
      const result = translateFiltersJSON(
        null as unknown as JSONArray,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should return empty array when filtersJSON is undefined', () => {
      const result = translateFiltersJSON(
        undefined as unknown as JSONArray,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should return empty array when filtersJSON is false', () => {
      const result = translateFiltersJSON(
        false as unknown as JSONArray,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should return empty array when filtersJSON is 0', () => {
      const result = translateFiltersJSON(
        0 as unknown as JSONArray,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should return empty array when filtersJSON is empty string', () => {
      const result = translateFiltersJSON(
        '' as unknown as JSONArray,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should translate empty array to empty array', () => {
      const result = translateFiltersJSON(
        [],
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(result.success).toBe(true);
      expect(getSuccessData(result)).toEqual([]);
    });

    it('should return error for array of strings instead of function calls', () => {
      const invalidFiltersJSON = ['DM.Country.Country', 'DM.Brand.Brand'] as unknown as JSONArray;

      const result = translateFiltersJSON(
        invalidFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)).toContain(
        'Invalid filters JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should return error for array of objects missing function property', () => {
      const invalidFiltersJSON = [
        { args: ['DM.Country.Country', ['United States']] },
      ] as unknown as JSONArray;

      const result = translateFiltersJSON(
        invalidFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)).toContain(
        'Invalid filters JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should return error for array of objects missing args property', () => {
      const invalidFiltersJSON = [{ function: 'filterFactory.members' }] as unknown as JSONArray;

      const result = translateFiltersJSON(
        invalidFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)).toContain(
        'Invalid filters JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should return error for array containing non-objects', () => {
      const invalidFiltersJSON = [
        { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
        'not an object',
      ] as unknown as JSONArray;

      const result = translateFiltersJSON(
        invalidFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)).toContain(
        'Invalid filters JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should return error for array containing null values', () => {
      const invalidFiltersJSON = [
        { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
        null,
      ] as unknown as JSONArray;

      const result = translateFiltersJSON(
        invalidFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)).toContain(
        'Invalid filters JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should translate valid parsed function call array', () => {
      const validFiltersJSON = [
        { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
      ];

      const result = translateFiltersJSON(
        validFiltersJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(true);
      const data = getSuccessData(result);
      expect(Array.isArray(data)).toBe(true);
      expect(data).toHaveLength(1);
    });
  });

  describe('translateHighlightsFromJSONFunctionCall', () => {
    it('should translate highlights from parsed function calls', () => {
      const mockHighlightsJSON: ParsedFunctionCall[] = [
        { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
        { function: 'filterFactory.members', args: ['DM.Brand.Brand', ['Brand A']] },
      ];

      const result = translateHighlightsFromJSONFunctionCall(
        mockHighlightsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(true);
      const data = getSuccessData(result);
      expect(data).toHaveLength(2);
      expect(withoutGuids(data)).toMatchSnapshot();
    });

    it('should return error when trying to use measure function as highlight', () => {
      const mockHighlightsJSON: ParsedFunctionCall[] = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue'] },
      ];

      const result = translateHighlightsFromJSONFunctionCall(
        mockHighlightsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)[0]).toContain('Invalid filter JSON');
    });

    it('should return error when trying to use filter relation as highlight', () => {
      const mockHighlightsJSON: ParsedFunctionCall[] = [
        {
          function: 'filterFactory.logic.and',
          args: [
            {
              function: 'filterFactory.members',
              args: ['DM.Country.Country', ['United States']],
            },
            {
              function: 'filterFactory.members',
              args: ['DM.Brand.Brand', ['Brand A']],
            },
          ],
        },
      ];

      const result = translateHighlightsFromJSONFunctionCall(
        mockHighlightsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)[0]).toContain('Invalid filter JSON');
    });

    it('should return error when trying to use average measure as highlight', () => {
      const mockHighlightsJSON: ParsedFunctionCall[] = [
        { function: 'measureFactory.average', args: ['DM.Commerce.Revenue'] },
      ];

      const result = translateHighlightsFromJSONFunctionCall(
        mockHighlightsJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(result.success).toBe(false);
      expect(getErrors(result)[0]).toContain('Invalid filter JSON');
    });
  });
});
