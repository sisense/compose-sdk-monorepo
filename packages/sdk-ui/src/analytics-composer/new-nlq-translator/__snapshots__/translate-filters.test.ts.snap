// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translateFilters > should translate filter relations 1`] = `
{
  "composeCode": "filterFactory.logic.and(filterFactory.logic.or(filterFactory.members(DM.Country.Country, ['United States', 'Canada']), filterFactory.members(DM.Brand.Brand, ['Brand A', 'Brand B'])), filterFactory.measureGreaterThan(measureFactory.sum(DM.Commerce.Revenue), 1000))",
  "left": {
    "composeCode": "filterFactory.logic.or(filterFactory.members(DM.Country.Country, ['United States', 'Canada']), filterFactory.members(DM.Brand.Brand, ['Brand A', 'Brand B']))",
    "left": {
      "__serializable": "MembersFilter",
      "_name": "filter",
      "attribute": {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Country.Country",
        "dataSource": {
          "address": "LocalHost",
          "id": "localhost_aSampleIAAaECommerce",
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Country.Country]",
        "name": "Country",
        "type": "text-attribute",
      },
      "composeCode": "filterFactory.members(DM.Country.Country, ['United States', 'Canada'])",
      "config": {
        "deactivatedMembers": [],
        "disabled": false,
        "enableMultiSelection": true,
        "excludeMembers": false,
        "locked": false,
      },
      "description": "",
      "filterType": "members",
      "isScope": true,
      "members": [
        "United States",
        "Canada",
      ],
      "type": "filter",
    },
    "operator": "OR",
    "right": {
      "__serializable": "MembersFilter",
      "_name": "filter",
      "attribute": {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Brand.Brand",
        "dataSource": {
          "address": "LocalHost",
          "id": "localhost_aSampleIAAaECommerce",
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Brand.Brand]",
        "name": "Brand",
        "type": "text-attribute",
      },
      "composeCode": "filterFactory.members(DM.Brand.Brand, ['Brand A', 'Brand B'])",
      "config": {
        "deactivatedMembers": [],
        "disabled": false,
        "enableMultiSelection": true,
        "excludeMembers": false,
        "locked": false,
      },
      "description": "",
      "filterType": "members",
      "isScope": true,
      "members": [
        "Brand A",
        "Brand B",
      ],
      "type": "filter",
    },
  },
  "operator": "AND",
  "right": {
    "__serializable": "MeasureFilter",
    "_name": "filter",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Commerce.Revenue",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Commerce.Revenue]",
      "name": "Revenue",
      "type": "numeric-attribute",
    },
    "composeCode": "filterFactory.measureGreaterThan(measureFactory.sum(DM.Commerce.Revenue), 1000)",
    "config": {
      "disabled": false,
      "locked": false,
    },
    "description": "",
    "filterType": "measure",
    "isScope": true,
    "measure": {
      "__serializable": "DimensionalBaseMeasure",
      "aggregation": "sum",
      "attribute": {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Revenue",
        "dataSource": {
          "address": "LocalHost",
          "id": "localhost_aSampleIAAaECommerce",
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Revenue]",
        "name": "Revenue",
        "type": "numeric-attribute",
      },
      "composeCode": "measureFactory.sum(DM.Commerce.Revenue)",
      "dataSource": undefined,
      "description": "",
      "name": "sum Revenue",
      "sort": 0,
      "type": "basemeasure",
    },
    "operatorA": "fromNotEqual",
    "type": "filter",
    "valueA": 1000,
  },
}
`;

exports[`translateFilters > should translate filters 1`] = `
[
  {
    "__serializable": "MembersFilter",
    "_name": "filter",
    "attribute": {
      "__serializable": "DimensionalLevelAttribute",
      "composeCode": "DM.Commerce.Date.Years",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Commerce.Date]",
      "format": "yyyy",
      "granularity": "Years",
      "name": "Date",
      "type": "datelevel",
    },
    "composeCode": "filterFactory.members(DM.Commerce.Date.Years, ['2024-01-01T00:00:00', '2025-01-01T00:00:00'], { disabled: false, locked: false, excludeMembers: false, enableMultiSelection: true, deactivatedMembers: [] })",
    "config": {
      "deactivatedMembers": [],
      "disabled": false,
      "enableMultiSelection": true,
      "excludeMembers": false,
      "locked": false,
    },
    "description": "",
    "filterType": "members",
    "isScope": true,
    "members": [
      "2024-01-01T00:00:00",
      "2025-01-01T00:00:00",
    ],
    "type": "filter",
  },
  {
    "__serializable": "MembersFilter",
    "_name": "filter",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Country.Country",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Country.Country]",
      "name": "Country",
      "type": "text-attribute",
    },
    "composeCode": "filterFactory.members(DM.Country.Country, ['United States'])",
    "config": {
      "deactivatedMembers": [],
      "disabled": false,
      "enableMultiSelection": true,
      "excludeMembers": false,
      "locked": false,
    },
    "description": "",
    "filterType": "members",
    "isScope": true,
    "members": [
      "United States",
    ],
    "type": "filter",
  },
]
`;

exports[`translateFilters > translateHighlightsFromJSONFunctionCall > should translate highlights from parsed function calls 1`] = `
[
  {
    "__serializable": "MembersFilter",
    "_name": "filter",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Country.Country",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Country.Country]",
      "name": "Country",
      "type": "text-attribute",
    },
    "composeCode": "filterFactory.members(DM.Country.Country, ['United States'])",
    "config": {
      "deactivatedMembers": [],
      "disabled": false,
      "enableMultiSelection": true,
      "excludeMembers": false,
      "locked": false,
    },
    "description": "",
    "filterType": "members",
    "isScope": true,
    "members": [
      "United States",
    ],
    "type": "filter",
  },
  {
    "__serializable": "MembersFilter",
    "_name": "filter",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Brand.Brand",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Brand.Brand]",
      "name": "Brand",
      "type": "text-attribute",
    },
    "composeCode": "filterFactory.members(DM.Brand.Brand, ['Brand A'])",
    "config": {
      "deactivatedMembers": [],
      "disabled": false,
      "enableMultiSelection": true,
      "excludeMembers": false,
      "locked": false,
    },
    "description": "",
    "filterType": "members",
    "isScope": true,
    "members": [
      "Brand A",
    ],
    "type": "filter",
  },
]
`;
